File: typescript-examples/playwright-basics-ts/api/03-locators-and-selectors.ts
import { BrowserContext, Page } from "playwright";

/**
 * Locators and Selectors
 *
 * Demonstrates different ways to find elements:
 * - CSS selectors
 * - XPath selectors
 * - Built-in semantic locators (getByRole, getByText, etc.)
 * - Chaining and filtering
 */

interface Params {}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
) {
  await page.goto("https://books.toscrape.com/");
  await page.waitForLoadState("networkidle");

  // CSS Selectors
  const cssLocator = page.locator(".product_pod h3 a");
  const cssCount = await cssLocator.count();

  // XPath Selectors
  const xpathLocator = page.locator("//article[@class='product_pod']//h3/a");
  const xpathCount = await xpathLocator.count();

  // Built-in semantic locators
  // getByRole - find by ARIA role
  const nextButton = page.getByRole("link", { name: "next" });
  const hasNextButton = await nextButton.isVisible();

  // getByText - find by visible text
  const travelCategory = page.getByText("Travel", { exact: true });
  const hasTravelCategory = await travelCategory.isVisible();

  // Chaining and filtering
  const firstProduct = page.locator(".product_pod").first();
  const firstProductTitle = await firstProduct.locator("h3 a").getAttribute("title");

  // Using nth() to get specific element
  const thirdProduct = page.locator(".product_pod").nth(2);
  const thirdProductTitle = await thirdProduct.locator("h3 a").getAttribute("title");

  // Filter locators
  const inStockProducts = page.locator(".product_pod").filter({
    has: page.locator(".instock"),
  });
  const inStockCount = await inStockProducts.count();

  return {
    cssCount,
    xpathCount,
    hasNextButton,
    hasTravelCategory,
    firstProductTitle,
    thirdProductTitle,
    inStockCount,
  };
}

File: python-examples/playwright-python/api/03-locators-and-selectors.py
"""
Locators and Selectors

Demonstrates different ways to find elements:
- CSS selectors
- XPath selectors
- Built-in semantic locators (get_by_role, get_by_text, etc.)
- Chaining and filtering
"""

from playwright.async_api import Page
from typing import TypedDict


class Params(TypedDict):
    pass


async def automation(page: Page, params: Params | None = None, **_kwargs):
    await page.goto("https://books.toscrape.com/")
    await page.wait_for_load_state("networkidle")

    # CSS Selectors
    css_locator = page.locator(".product_pod h3 a")
    css_count = await css_locator.count()

    # XPath Selectors
    xpath_locator = page.locator("//article[@class='product_pod']//h3/a")
    xpath_count = await xpath_locator.count()

    # Built-in semantic locators
    # get_by_role - find by ARIA role
    next_button = page.get_by_role("link", name="next")
    has_next_button = await next_button.is_visible()

    # get_by_text - find by visible text
    travel_category = page.get_by_text("Travel", exact=True)
    has_travel_category = await travel_category.is_visible()

    # Chaining and filtering
    first_product = page.locator(".product_pod").first
    first_product_title = await first_product.locator("h3 a").get_attribute("title")

    # Using nth() to get specific element
    third_product = page.locator(".product_pod").nth(2)
    third_product_title = await third_product.locator("h3 a").get_attribute("title")

    # Filter locators
    in_stock_products = page.locator(".product_pod").filter(has=page.locator(".instock"))
    in_stock_count = await in_stock_products.count()

    return {
        "cssCount": css_count,
        "xpathCount": xpath_count,
        "hasNextButton": has_next_button,
        "hasTravelCategory": has_travel_category,
        "firstProductTitle": first_product_title,
        "thirdProductTitle": third_product_title,
        "inStockCount": in_stock_count,
    }
